File: app/api/schemas/rule.py
"""Pydantic response schemas for the rules API — mirrors rule_summary."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class RuleSummary(BaseModel):
    """One row from the rule_summary view. from_attributes lets this be
    built directly from a SQLAlchemy Row/mapping without manual field-by-field
    construction."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    customer_id: int
    qradar_rule_id: int
    identifier: str | None
    name: str | None
    object_type: str
    building_block_subtype: str | None
    type: str | None
    enabled: bool | None
    owner: str | None
    origin: str | None
    created_at: datetime | None
    updated_at: datetime | None
    last_event_at: datetime | None
    last_event_count: int | None
    last_offense_id: str | None
    tactics: list[str]
    techniques: list[str]
    sub_techniques: list[str]


class RuleListResponse(BaseModel):
    count: int
    results: list[RuleSummary]


class RuleHealthMetrics(BaseModel):
    total_rules: int
    total_building_blocks: int
    enabled_rules: int
    disabled_rules: int
    enabled_triggered: int
    enabled_not_triggered: int

File: frontend/src/lib/types/rule.ts
export interface Rule {
	id: number;
	customer_id: number;
	qradar_rule_id: number;
	identifier: string | null;
	name: string | null;
	object_type: 'RULE' | 'BUILDING_BLOCK';
	building_block_subtype: string | null;
	type: string | null;
	enabled: boolean | null;
	owner: string | null;
	origin: string | null;
	created_at: string | null;
	updated_at: string | null;
	last_event_at: string | null;
	last_event_count: number | null;
	last_offense_id: string | null;
	tactics: string[];
	techniques: string[];
	sub_techniques: string[];
}

export interface RulesResponse {
	count: number;
	results: Rule[];
}

export interface RuleHealthMetrics {
	total_rules: number;
	total_building_blocks: number;
	enabled_rules: number;
	disabled_rules: number;
	enabled_triggered: number;
	enabled_not_triggered: number;
}